File: learn/graph/bellman-ford/bellman_ford.go


import "math"

type BFEdge struct {
	u      int
	v      int
	weight int
}

type BellmanFord struct {
	vertices int
	edges    []BFEdge
}

func NewBellmanFord(v int) *BellmanFord {
	return &BellmanFord{
		vertices: v,
		edges:    []BFEdge{},
	}
}

func (bf *BellmanFord) AddEdge(u, v, weight int) {
	bf.edges = append(bf.edges, BFEdge{u, v, weight})
}

func (bf *BellmanFord) ShortestPath(start int) ([]int, bool) {
	dist := make([]int, bf.vertices)
	for i := range dist {
		dist[i] = math.MaxInt32
	}
	dist[start] = 0

	for i := 1; i < bf.vertices; i++ {
		for _, edge := range bf.edges {
			if dist[edge.u] != math.MaxInt32 && dist[edge.u]+edge.weight < dist[edge.v] {
				dist[edge.v] = dist[edge.u] + edge.weight
			}
		}
	}

	for _, edge := range bf.edges {
		if dist[edge.u] != math.MaxInt32 && dist[edge.u]+edge.weight < dist[edge.v] {
			return nil, false // Negative cycle detected
		}
	}

	return dist, true
}

File: learn/graph/bellman-ford/bellman_ford.cpp
struct Edge {
    int source, target, weight;
};

class BellmanFord {
private:
    int V;
    vector<Edge> edges;
    const int INF = 1e9;

public:
    BellmanFord(int vertices) {
        V = vertices;
    }

    void addEdge(int u, int v, int weight) {
        edges.push_back({u, v, weight});
    }

    // Returns a distance array, or an empty vector if negative cycle is detected
    vector<int> shortestPath(int start) {
        vector<int> dist(V, INF);
        dist[start] = 0;

        for (int i = 0; i < V - 1; ++i) {
            for (auto edge : edges) {
                int u = edge.source;
                int v = edge.target;
                int weight = edge.weight;

                if (dist[u] != INF && dist[u] + weight < dist[v]) {
                    dist[v] = dist[u] + weight;
                }
            }
        }

        for (auto edge : edges) {
            int u = edge.source;
            int v = edge.target;
            int weight = edge.weight;

            if (dist[u] != INF && dist[u] + weight < dist[v]) {
                return {}; // Negative weight cycle
            }
        }

        return dist;
    }
};